### PR TITLE
TFS 798892:[Zurich] TN3270 custom platform sample

### PR DESCRIPTION
**Pull request by StephanieZ**
Ready for master.

### CODE REVIEW

File: SampleScripts/GenericRacfTn3270.json [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+      { "FuncUserName": { "Type": "String", "Required": true } },
+      { "FuncPassword": { "Type": "Secret", "Required": true } },
+      { "Port": { "Type": "Integer", "Required": false, "DefaultValue": 23 } },
+      { "Timeout": { "Type": "Integer", "Required": false, "DefaultValue": 30 } },
+      { "UseSsl": { "Type": "boolean", "Required": false, "DefaultValue": true } },
+      { "SkipServerCertValidation": { "Type": "boolean", "Required": false, "DefaultValue": false } },
+      { "WorkstationId": { "Type": "String", "Required": false, "DefaultValue": "" } }
+    ],
+    "Do": [
+      { "SetItem": { "Name": "FuncUserName", "Value": "%{ FuncUserName.ToUpper() }%" } },
```
Comment (bdahneke) [inline SampleScripts/GenericRacfTn3270.json]: As an example script this behavior may be confusing. We are setting FuncUserName prior to calling LoginFuncRacf. We are not passing it as a parameter rather we are depending on the behavior of variables set at the top level being created as globals. I think it would be better to pass FuncUserName as a parameter.

File: SampleScripts/GenericRacfTn3270.json [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+      { "AccountPassword": { "Type": "Secret", "Required": true } },
+      { "Port": { "Type": "Integer", "Required": false, "DefaultValue": 23 } },
+      { "Timeout": { "Type": "Integer", "Required": false, "DefaultValue": 30 } },
+      { "UseSsl": { "Type": "boolean", "Required": false, "DefaultValue": true } },
+      { "SkipServerCertValidation": { "Type": "boolean", "Required": false, "DefaultValue": false } },
+      { "WorkstationId": { "Type": "String", "Required": false, "DefaultValue": "" } }
+    ],
+    "Do": [
+      { "Status": { "Type": "Checking", "Percent": 20, "Message": { "Name": "VerifyingPassword" } } },
+      { "SetItem": { "Name": "AccountUserName", "Value": "%{ AccountUserName.ToUpper() }%" } },
```
Comment (bdahneke) [inline SampleScripts/GenericRacfTn3270.json]: Again, pass these as parameters.

File: SampleScripts/GenericRacfTn3270.json [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+      { "SetItem": {"Name": "FuncUserName", "Value": "%{ FuncUserName.ToUpper() }%" } },
+      { "Function": { "Name": "LoginFuncRacf", "ResultVariable": "LoginResult" } },
+      { "Condition": { "If": "!LoginResult", "Then": { "Do": [ { "Throw": { "Value": "Account %FuncUserName% not found" } } ] } } },
+      { "Function": { "Name": "LogoutRacf" } },
+      { "Function": { "Name": "LoginUserRacf", "ResultVariable": "LoginResult" } },
+      { "Condition": { "If": "!LoginResult", "Then": { "Do": [ { "Return": { "Value": "False" } } ] } } },
+      { "Function": { "Name": "LogoutRacf" } },
+      { "Return": { "Value": "True" } }
```
Comment (bdahneke) [inline SampleScripts/GenericRacfTn3270.json]: Bool is better than a string of the bool here

File: SampleScripts/GenericRacfTn3270.json [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+      },
+      { "Send": { "ConnectionObjectName": "Tn3270Connection", "Buffer": "ALU %AccountUserName% PASSWORD(%NewPassword%) NOEXPIRED\n\f","ContainsSecret": true } },
+      { "Receive": { "ConnectionObjectName": "Tn3270Connection", "BufferName": "AluCmdBuffer", "ContainsSecret": true } },
+      { "SetItem": { "Name": "Global:PwdChangeResult", "Value": "false" } },
+      { "SetItem": { "Name": "Global:InvalidPassword", "Value": false } },
+      {
+        "Condition": {
+          "If": "Regex.IsMatch(AluCmdBuffer,@\"(.*READY[ \n]*$)\")",
+          "Then": {
+            "Do": [ { "SetItem": { "Name": "PwdChangeResult", "Value": "true" } } ]
```
Comment (bdahneke) [inline SampleScripts/GenericRacfTn3270.json]: consider changing PwdChangeResult to a bool

File: SampleScripts/GenericRacfTn3270.json [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+      { "FuncUserName": { "Type": "String", "Required": true } },
+      { "FuncPassword": { "Type": "Secret", "Required": true } },
+      { "Port": { "Type": "Integer", "Required": false, "DefaultValue": 23 } },
+      { "Timeout": { "Type": "Integer", "Required": false, "DefaultValue": 30 } },
+      { "UseSsl": { "Type": "boolean", "Required": false, "DefaultValue": true } },
+      { "SkipServerCertValidation": { "Type": "boolean", "Required": false, "DefaultValue": false } },
+      { "WorkstationId": { "Type": "String", "Required": false, "DefaultValue": "" } }
+    ],
+    "Do": [
+      { "Function": { "Name": "LoginFuncRacf", "ResultVariable": "LoginResult" }, "Parameters": ["%FuncUserName%"] },
```
Comment (NathanTurnbow) [inline SampleScripts/GenericRacfTn3270.json]: This "Parameters" is inside the wrong curly brace, it's not really parsing.

File: SampleScripts/GenericRacfTn3270.json [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+      { "UseSsl": { "Type": "boolean", "Required": false, "DefaultValue": true } },
+      { "SkipServerCertValidation": { "Type": "boolean", "Required": false, "DefaultValue": false } },
+      { "WorkstationId": { "Type": "String", "Required": false, "DefaultValue": "" } }
+    ],
+    "Do": [
+      { "Function": { "Name": "LoginFuncRacf", "ResultVariable": "LoginResult" } },
+      { "Condition": {
+          "If": "!LoginResult",
+          "Then": {
+            "Do": [ { "Return": { "Value": true } } ]
```
Comment (NathanTurnbow) [inline SampleScripts/GenericRacfTn3270.json]: Should it return false if we fail to login?

Comment (NathanTurnbow) [reply SampleScripts/GenericRacfTn3270.json]: In check password we actually throw and give a detailed message, that seems better even than a failure boolean with no details.

Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: yep, good catch

File: SampleScripts/GenericRacfTn3270.json [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+          "Type": "Changing",
+          "Message": {"Name": "ChangingPassword",
+            "Parameters": [ "%AccountUserName%" ] },
+          "Percent": "50"
+        }
+      },
+      { "Send": { "ConnectionObjectName": "Tn3270Connection", "Buffer": "ALU %AccountUserName% PASSWORD(%NewPassword%) NOEXPIRED\n\f","ContainsSecret": true } },
+      { "Receive": { "ConnectionObjectName": "Tn3270Connection", "BufferName": "AluCmdBuffer", "ContainsSecret": true } },
+      { "SetItem": { "Name": "Global:PwdChangeResult", "Value": false } },
+      { "SetItem": { "Name": "Global:InvalidPassword", "Value": false } },
```
Comment (NathanTurnbow) [inline SampleScripts/GenericRacfTn3270.json]: This doesn't look like it's used

File: SampleScripts/GenericRacfTn3270.json [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+      { "Status": {
+          "Type": "Changing",
+          "Message": {"Name": "ChangingPassword",
+            "Parameters": [ "%AccountUserName%" ] },
+          "Percent": "50"
+        }
+      },
+      { "Send": { "ConnectionObjectName": "Tn3270Connection", "Buffer": "ALU %AccountUserName% PASSWORD(%NewPassword%) NOEXPIRED\n\f","ContainsSecret": true } },
+      { "Receive": { "ConnectionObjectName": "Tn3270Connection", "BufferName": "AluCmdBuffer", "ContainsSecret": true } },
+      { "SetItem": { "Name": "Global:PwdChangeResult", "Value": false } },
```
Comment (NathanTurnbow) [inline SampleScripts/GenericRacfTn3270.json]: No need for global here, everything will work the same without it

File: SampleScripts/GenericRacfTn3270.json (hunk 7 above)
Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: yeah I removed it already.

File: SampleScripts/GenericRacfTn3270.json (hunk 8 above)
Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: agreed

File: SampleScripts/GenericRacfTn3270.json [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+      { "Condition": { "If": "!LoginResult", "Then": { "Do": [ { "Return": { "Value": false } } ] } } },
+      { "Status": {
+          "Type": "Changing",
+          "Message": {"Name": "ChangingPassword",
+            "Parameters": [ "%AccountUserName%" ] },
+          "Percent": "50"
+        }
+      },
+      { "Send": { "ConnectionObjectName": "Tn3270Connection", "Buffer": "ALU %AccountUserName% PASSWORD(%NewPassword%) NOEXPIRED\n\f","ContainsSecret": true } },
+      { "Receive": { "ConnectionObjectName": "Tn3270Connection", "BufferName": "AluCmdBuffer", "ContainsSecret": true } },
```
Comment (NathanTurnbow) [inline SampleScripts/GenericRacfTn3270.json]: Do we know that RACF reveals passwords in the response? Or is the ContainsSecret preventing useful log output without a reason?

Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: Yes it definitely leaks passwords. The buffer doesn't get cleared on each Receive, so it keeps resending the password until we disconnect. I have another bug assigned to me to go back and clear the buffer and remove this "ContainsSecret = true" and a few others in TN3270 platforms. that's outside the scope of this TFS.

Comment (NathanTurnbow) [reply SampleScripts/GenericRacfTn3270.json]: But this is going to be sent to Zurich, right? Wouldn't it be better to fix it and then send them something without the bandaid bug fix?
Also, does that mean that everywhere in here in all of the receive's we're looking through the *full history* of the whole session rather than just a response from the previous Send? Is that going to mess up the ExpectRegex anywhere?

Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: It doesn't mess up ExpectRegex, because we only withhold the buffer content from the logs, it's not like we throw it out completely. So ExpectRegex can still "see" it and use it. I'll leave the other question to @DanPeterson. I don't see the difference between having ContainsSecret on the Send versus the Receive. If we are going to remove one, why not remove all of them if Zurich needs to see what's going on, they might as well see all ofit.

File: SampleScripts/GenericRacfTn3270.json [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+                      { "Function": { "Name": "LogoutRacf" } },
+                      { "Return": { "Value": false } }
+                    ]
+                  }
+                }
+              }
+            ]
+          },
+          {
+            "Name": "LoginUserRacf",
```
Comment (NathanTurnbow) [inline SampleScripts/GenericRacfTn3270.json]: If this function is identical to the function above it with the exception of Account username/password vs Func username/password, I would say pass in parameters "username" and "password" and consolidate the two.

Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: It's not identical, they use different statuses.

File: SampleScripts/GenericRacfTn3270.json (hunk 9 above)
Comment (NathanTurnbow) [reply SampleScripts/GenericRacfTn3270.json]: @StephanieZ So here's what I'm trying to say. The way I understand what you said is that *every time* we do a receive, we get the whole screen - like the whole history of everything on the buffer, not just what's on the buffer since last we received. There is a ton of logic in here that does stuff like `Regex.IsMatch(<somethingWeJustReceived>,"RegexLookingForAStaticString")`. If that regex is comparing against the *entire history* what's going to happen is you will get false positives everywhere for the regex match. For example, in CheckSystem line 40 there is `Regex.IsMatch(LuCmdBuffer, $\"^(.*USER={FuncUserName}.*)\")` which happens after a login/logout using the same leftover connection object `Tn3270Connection` which is a global. What is that '^' going to match against? I'm thinking it's the beginning of the whole buffer, but clearly that's not the intent of that line. In other words, it's never going to evaluate to true, so we're going to go on and say that checkSystem was successful, when really we should be saying `InsufficientPrivilegesAccessPassword`.

So basically by not figuring out a fix for clearing out the buffer we have holes in this script and logic that doesn't really work beyond just missing some debug details

Comment (StephanieZ) [reply SampleScripts/GenericRacfTn3270.json]: Ok I misunderstood what you were saying the issue was. I will start fixing the buffer issue first.

File: SampleScripts/GenericRacfTn3270.json (hunk 10 above)
Comment (NathanTurnbow) [reply SampleScripts/GenericRacfTn3270.json]: Statuses could also be refactored around a boolean. What consolidating these two functions would do is save them the headache if they find a bug in this later of needing to know to fix it in two places

**Comment by StephanieZ** [comment]
closing this until I get RACF rewritten.
